### ISSUE TITLE
Network connection always required to start profiling

### ISSUE THREAD

**Issue by batzen** [JetBrains/profiler-self-api#13]
Inside `DownloadAsync` in `PrerequisiteBase.cs` there is a call to `TryGetRunner`.
But `TryGetRunner` does not include subdirectories and versions when checking.
That's only done by `TryGetRunnerEx`.
Because of that NuGet is always contacted, even though it does not end up downloading something if tools were already downloaded.

To fix this issue `DownloadAsync`  should also call `TryGetRunnerEx`.
Do you accept PRs?

**Comment by gassanosh** [comment]
Hi! First let me explain why do we need to have access to nuget.org.
The self API provides two specific functionalities for obtaining Command Line Tools (clt):
1. The self-api keeps the clt up to date itself. Here we need to check nuget.org in case bugfix were released. In that mode it is expected that the tool will check nuget.org, but as you said the traffic should be minimized. It's a default mode.
2. For cases you don't have access to the Internet there is an option to specify a folder with clt. For that case you must be sure that the compatible version was provided and you need to manually update clt in case critical fixes were released.

So in other words it is expected that by default the self-api will check nuget.org to make sure the latest version of clt is used.
If you don't want it, just put compatible clt into any folder and specify it as downloadTo argument for EnsurePrerequisite.

**Comment by batzen** [comment]
Putting it locally to prevent the call to nuget is exactly what we try to do.
But copying the exact same directory layout that's present after downloading through nuget does not work because the checked directory is not correct.

**Comment by gassanosh** [comment]
You shouldn't copy self-api internal directory layout, in that case it will work just as described in first case.
If you want to make it use pre-downloaded version of clt and disable auto update just specify a folder where dottrace binary could be found. In xml doc it is the case 1:
```
   /// <summary>
    /// Makes sure that the dotTrace command-line profiler is downloaded and is ready to use.
    /// </summary>
    /// <remarks>
    /// 1. Looks for dotTrace executable in the same directory with the running assembly. Uses it if it's found.<br/>
    /// 2. Downloads `JetBrains.dotTrace.CommandLineTools` NuGet package into the <paramref name="downloadTo"/>
    /// directory and uses the dotTrace command-line profiler from this package. The package version is defined by <see cref="CommandLineToolsConfig.NupkgVersion"/>.
    /// The command-line profiler is saved to `{downloadTo}/dotTrace.{NupkgVersion}`
    /// If the executable file exists, a new one is not downloaded.
    /// </remarks>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <param name="progress">Download progress callback from 0.0 to 100.0. If null, progress is not reported.</param>
    /// <param name="nugetUrl">URL of NuGet mirror. If null, www.nuget.org is used.</param>
    /// <param name="nugetApi">NuGet API version.</param>
    /// <param name="downloadTo">NuGet download destination directory. If null, %LocalAppData% is used.</param>
    public static Task EnsurePrerequisiteAsync(
      CancellationToken cancellationToken,
      IProgress<double> progress = null,
      Uri nugetUrl = null,
      NuGetApi nugetApi = NuGetApi.V3,
      string downloadTo = null)
```

**Comment by batzen** [comment]
It's an easy fix to make the checks the same and prevent confusion when trying to do an offline deployment.
It took me quite some time to find out why it didn't work in an offline deployment.
That's why I asked if PRs are accepted.

**Comment by gassanosh** [comment]
Thank you for the suggestion.
However, as I said before, this change will unintentionally break the default behavior of the api.
In this case, it might be better to stick with the current implementation since the requested feature (skipping downloading of clt) is already available in the api.
Probably we could add new method/overload with more clear description or more clear describe this case in xmldoc. If you have a specific idea for the new method name or XML documentation improvement, you could mention it to provide more clarity.
But I'm afraid we cannot afford breaking default behavior.

**Comment by batzen** [comment]
Now i get it.
Would it be ok if i create a PR with changes that would solve the issue for us in a way that won't break the current behavior?

**Comment by gassanosh** [comment]
As I said before the problem is only in method naming. Now it's not obvious what to call when it worked in the environment without the Internet. We will fix the issue.
But right now you still could use EnsurePrerequisite for "offline" scenarios.
For example you could call:
```
    DotTrace.EnsurePrerequisite(nugetUrl: new Uri("about:none"), downloadTo: "<path_to_clt>");
```
Just make sure the folder <path_to_clt> contains command line tools binaries in its root. Don't put unpacked binaries to the nested folders with version name. In other words `ls <path_to_clt>` (or `dir <path_to_clt>`) should look something like this:
```
....
JetBrains...
dottrace
dottrace.exe
....
```
